packages/core/contracts/zodiac/OptimisticOracleModule.sol

```diff
@@ -0,0 +1,345 @@
+        require(_getCollateralWhitelist().isOnWhitelist(address(_collateral)), "bond token not supported");
+        collateral = _collateral;
+    }
+
+    function setRules(string memory _rules) public onlyOwner {
+        // Set reference to the rules for the avatar (e.g. an IPFS hash or URI).
+        rules = _rules;
+    }
+
+    function setLiveness(uint64 _liveness) public onlyOwner {
+        // Set liveness for disputing proposed transactions.
+        require(_liveness > 0, "liveness can't be 0");
+        liveness = _liveness;
+    }
+
+    function setIdentifier(bytes32 _identifier) public onlyOwner {
+        // Set identifier which is used along with the rules to determine if transactions are valid.
+        require(_getIdentifierWhitelist().isIdentifierSupported(_identifier), "identifier not supported");
+        identifier = _identifier;
+    }
+
+    function sync() public nonReentrant {
+        // Sync the store and oracle contract addresses as well as the final fee.
+        _sync();
+    }
+
+    function proposeTransactions(Transaction[] memory _transactions, bytes memory _explanation) public nonReentrant {
+        // note: Optional explanation explains the intent of the transactions to make comprehension easier.
+        uint256 id = prevProposalId + 1;
+        uint256 time = getCurrentTime();
+        address proposer = msg.sender;
+
+        // Create proposal in memory to emit in an event.
+        Proposal memory proposal;
+        proposal.requestTime = time;
+
+        // Construct the ancillary data.
+        bytes memory ancillaryData = _explanation;
+        proposal.explanation = _explanation;
+        AncillaryData.appendKeyValueUint(ancillaryData, "id", id);
+        AncillaryData.appendKeyValueAddress(ancillaryData, "module", address(this));
+
+        // Add transactions to proposal in memory.
+        for (uint256 i = 0; i < _transactions.length; i++) {
+            require(_transactions[i].to != address(0), "The `to` address cannot be 0x0");
+            // If the transaction has any data with it the recipient must be a contract, not an EOA.
+            if (_transactions[i].data.length > 0) {
+                require(_isContract(_transactions[i].to), "EOA can't accept tx with data");
+            }
+        }
+        proposal.transactions = _transactions;
+
+        // Add transaction data to the proposal data to be hashed and stored in the contract.
+        bytes memory proposalData = abi.encodePacked(ancillaryData);
+
+        for (uint256 i = 0; i < _transactions.length; i++) {
+            proposalData = bytes.concat(abi.encodePacked(_transactions[i].to));
+            proposalData = bytes.concat(abi.encodePacked(_transactions[i].value));
+            proposalData = bytes.concat(abi.encodePacked(_transactions[i].data));
+            proposalData = bytes.concat(abi.encodePacked(_transactions[i].operation));
+        }
+
+        proposalHashes[id] = keccak256(abi.encodePacked(proposalData));
```

Comment:
Cool, will make that change.

Reply:
Looks like this should work. Rather than encoding the whole proposal, though, I just encode the `_transactions`. This makes it easier to check during the execution step since you don't have to re-build the whole proposal in memory just to compare hashes.